File: epos_restaurant_2023/gym/report/training_attendance_report/training_attendance_report.py
# Copyright (c) 2024, Tes Pheakdey and contributors
# For license information, please see license.txt

import frappe
from datetime import datetime


def execute(filters=None):

	validate(filters)

	report_data = []
	skip_total_row=False
	message=None
	report_data = get_report_data(filters=filters)
 
	return get_columns(filters=filters), report_data, message, None, None,skip_total_row


def validate(filters):
	if not filters.class_type:
		filters.class_type = frappe.db.get_list("GYM Class Types",pluck='name')
	
	if not filters.time_training:
		filters.time_training = frappe.db.get_list("GYM Training Time",pluck='name')

def get_columns(filters):
	columns = []
	columns.append( {"label":"Code", "fieldname":"code","fieldtype":"Dynamic Link","options":"reference_doctype", "align":"left",'width':100})
	columns.append( {"label":"Member Type", "fieldname":"member_type","fieldtype":"Data", "align":"center",'width':140})
	columns.append( {"label":"Name En", "fieldname":"name_en","fieldtype":"Data", "align":"left",'width':170})
	columns.append( {"label":"Name Kh", "fieldname":"name_kh","fieldtype":"Data", "align":"left",'width':170})
	columns.append( {"label":"Gender", "fieldname":"gender","fieldtype":"Data", "align":"left",'width':80})
	columns.append( {"label":"Phone", "fieldname":"phone_number","fieldtype":"Data", "align":"left",'width':200})
	columns.append( {"label":"Training Date", "fieldname":"calendar_date","fieldtype":"Date", "align":"left",'width':120})
	columns.append( {"label":"Class", "fieldname":"class","fieldtype":"Data", "align":"left",'width':170})
	columns.append( {"label":"Time Training", "fieldname":"time_training","fieldtype":"Data", "align":"left",'width':170})
	columns.append( {"label":"Check In", "fieldname":"check_in_date","fieldtype":"Data", "align":"left",'width':210,})
	columns.append( {"label":"Check Out", "fieldname":"check_out_date","fieldtype":"Data", "align":"left",'width':210})

	return columns


def get_report_data(filters):
	sql = """select 
		`name`,
		if(member is null , trainer, member) as `code`,
		if(member is null, trainer_name_en, member_name ) as name_en,
		if(member is null, trainer_name_kh, member_name_kh ) as name_kh,
		if(member is null, gender, member_gender ) as gender,
		if(member is null,phone , concat(phone_number_1,'/',phone_number_2) ) as phone_number,
		reference_doctype,
		if(reference_doctype = 'Trainer','Trainer','Member') as member_type,
		class,
		time_training,
		calendar_date,
		check_in_date,
		check_out_date,	
		if(check_out_date is null, 0, 1) as is_check_out

	from `tabTraining Attendance` 
	where 1 = 1 
	and (calendar_date between %(start_date)s and %(end_date)s)
	and class in %(class_type)s 
	and time_training in %(time_training)s
	
	order by calendar_date"""

	result = frappe.db.sql(sql,filters, as_dict = 1)
	for row in result:
		if row.get("check_in_date"):
			row["check_in_date"] = row["check_in_date"].strftime("%Y-%m-%d %I:%M:%S %p") if isinstance(row["check_in_date"], datetime) else row["check_in_date"]
		if row.get("check_out_date"):
			row["check_out_date"] = row["check_out_date"].strftime("%Y-%m-%d %I:%M:%S %p") if isinstance(row["check_out_date"], datetime) else row["check_out_date"]

	return result

File: epos_restaurant_2023/gym/report/training_attendance_report/training_attendance_report.js
// Copyright (c) 2024, Tes Pheakdey and contributors
// For license information, please see license.txt

frappe.query_reports["Training Attendance Report"] = {
	onload: function(report) {
		report.page.add_inner_button("Preview Report", function () {
			frappe.query_report.refresh();
		});
		const custom_css = `
            <style>
                .datatable .dt-row .dt-cell:first-child .dt-cell__content{
					min-width: 100px !important;
					width: 100px !important;
					align-content: center !important;
				}
            </style>`;
        $(custom_css).appendTo("head");
	},
	
	"filters": [
		{
			"fieldname":"start_date",
			"label": __("Training (Start Date)"),
			"fieldtype": "Date",
			"default":frappe.datetime.get_today(),	
			"on_change": function (query_report) { },		 
			"reqd": 1
		},
		{
			"fieldname":"end_date",
			"label": __("Training (End Date)"),
			"fieldtype": "Date",
			"default":frappe.datetime.get_today(),
			"on_change": function (query_report) { },
			"reqd": 1
		},

		
		{
			"fieldname": "class_type",
			"label": __("Class Type"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				 // Fetch all the available training times
				 return frappe.db.get_list('GYM Class Types', {
					filters: {
						'name': ['like', '%' + txt + '%'] // Adjust this filter as necessary
					},
					fields: ['name', 'disabled'],
					order_by: 'name ASC'  // Sort based on 'sort_order' field
				}).then(function(results) {
					// Map the results to match the expected format
					return results.map(function(row) {
						let description = __("Actived");
						if(row.disabled == 1){
							description = __("Disabled");
						}
						return {
							value: row.name, // Or any field you want to display as the value
							description: description // Or another field for description if needed
						};
					});
				});
			},
			"on_change": function (query_report) { },
			
		},
		{
			"fieldname": "time_training",
			"label": __("Time Training"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				 // Fetch all the available training times
				 return frappe.db.get_list('GYM Training Time', {
					filters: {
						'name': ['like', '%' + txt + '%'] // Adjust this filter as necessary
					},
					fields: ['name', 'typical_time', 'sort_order'],
					order_by: 'sort_order ASC'  // Sort based on 'sort_order' field
				}).then(function(results) {
					// Map the results to match the expected format
					return results.map(function(row) {
						return {
							value: row.name, // Or any field you want to display as the value
							description: row.typical_time // Or another field for description if needed
						};
					});
				});
			},
			"on_change": function (query_report) { },
			
		},

	],

	"formatter": function(value, row, column, data, default_formatter) {
	
		value = default_formatter(value, row, column, data);

		if (data && data.is_group==1) {
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");
			

			value = $value.wrap("<p></p>").parent().html();
		}
		
		return value;
	},
};
